### DISCUSSION TITLE
How to get access to previous completed array?

### DISCUSSION THREAD

**Discussion by TLINDEN** [adoyle-h/bash-completor#2] [Q&A]
Hello,

first of all: i really like this script, it helps me a lot!

My problem is as follows: I use the following completion configuration (shortened):
```
output=_jaildk-completion.bash
cmd=jaildk
cmd_opts=()

subcmds=(rc)

reply_jail() {
    local jails=$(ls $JAILDIR/etc)
    COMPREPLY=( $(compgen -W "${jails[*]}" -- "$cur") )
}

modes='start,stop,status,restart'
#rcscripts=??

subcmd_opts_rc=(-m:$modes -r:@rcscripts)
subcmd_opts_rc=(-m:$modes -r)
subcmd_args_rc=@jail
```
So, the command `jaildk rc` executes a command (an rc script in fact) inside a jail. I get completion for the jails. However, in order to get completion for the available rc-scripts in a particular jail, I need to know which jail the user has selected by completion at this point:
```
jaildk rc dnsjail -m status -r<tab>
```
That is, when the user hits `<tab>` after `-r` AND a jail has already been selected like `dnsjail` in this case, then I need that name `dnsjail` to get a list of valid rc-scripts active in that jail.

Is something like this possible with bash-completor?

**Comment by adoyle-h** [comment]
The `COMP_WORDS` and `COMP_CWORD` variables are accessible in `reply_` functions. So you can try below codes.

```
reply_rcscripts() {
    local jail_type=${COMP_WORDS[2]}  # jail_type=dnsjail
    COMPREPLY=( $jail_type )
}
```

Please read the [Builtin gloabl variables](https://github.com/adoyle-h/bash-completor/blob/master/docs/syntax.md#builtin-gloabl-variables).

**Comment by TLINDEN** [comment]
Hi,

that works insofar, as I am now getting the expected list of matches, e.g.:
```
jaildk rc dns -r <tab><tab>
knotexporter   node_exporter
```

However, when I hit `k<tab>` then a literal TAB is being inserted, the completion doesn't work any longer. Seems `${COMP_WORDS[2]}` is only valid as long I did not enter anything.

The code looks as follows now:
```
reply_rcscripts() {
    local jail=${COMP_WORDS[2]}
    local scripts=$(jaildk _get_rc_scripts $jail | sed 's|.*/||') 
    COMPREPLY=( $(compgen -W "${scripts[*]}" -- "$cur") )
}
```

Oh and another weird behavior, when I enter directly (that is, no `<tab>` after `-r`:
```
jaildk rc dns -r k<tab>
```

then I get:
```
notexporter aildk rc dns -r k
```
So, a `\r` has been inserted somewhere?

**Reply by adoyle-h** [reply]
Sorry, I cannot reproduce your issues. I try the below script and it works well.

```bash
output=_jaildk-completion.bash
cmd=jaildk
cmd_opts=()

subcmds=(rc)

reply_jail() {
    COMPREPLY=( dns )
}

modes='start,stop,status,restart'

subcmd_opts_rc=(-m:$modes -r:@rcscripts)
subcmd_args_rc=@jail

reply_rcscripts() {
    local jail=${COMP_WORDS[2]}
    if [[ $jail == "dns" ]]; then
      COMPREPLY=( $(compgen -W "k1 k2 k3" -- "$cur") )
    else
      COMPREPLY=( "hello world" )
    fi
}
```

I think you should check the value of `local scripts`.

```bash
reply_rcscripts() {
    local jail=${COMP_WORDS[2]}
    local scripts=$(jaildk _get_rc_scripts $jail | sed 's|.*/||') 
    echo "[jail=$jail] [cur=$cur] [scripts=${scripts[*]}]" >> bash-completor-debug.log
    COMPREPLY=( $(compgen -W "${scripts[*]}" -- "$cur") )
}
```

**Comment by TLINDEN** [comment]
Hi,

based on your suggestion I now found a workaround. The problem seemed to be the way I fetched the candidates for the `@rcfiles` completion using `rcorder` et.al. The workaround is much simpler and now it works, even fast.


Thanks a lot!
Tom
